components/director/internal/domain/application/service.go

```diff
@@ -749,26 +760,100 @@ func (s *service) Merge(ctx context.Context, destID, srcID string) (*model.Appli
 		return nil, errors.Errorf("Could not determine status of source application with id %s", srcID)
 	}
 
-	if srcApp.Status.Condition == model.ApplicationStatusConditionConnected {
+	if srcApp.Status.Condition != model.ApplicationStatusConditionInitial {
 		return nil, errors.Errorf("Cannot merge application with id %s, because it is in a %s status", srcID, model.ApplicationStatusConditionConnected)
 	}
 
+	log.C(ctx).Infof("Merging applications with ids %s and %s", destID, srcID)
 	if err := mergo.Merge(destApp, *srcApp); err != nil {
 		return nil, errors.Wrapf(err, "while trying to merge applications with ids %s and %s", destID, srcID)
 	}
 
+	log.C(ctx).Infof("Merging labels for applications with ids %s and %s", destID, srcID)
+	destAppLabelsMerged, err := s.handleMergeLabels(srcAppLabels, destAppLabels)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while trying to merge labels for applications with ids %s and %s", destID, srcID)
+	}
+
 	log.C(ctx).Infof("Deleting source application with id %s", srcID)
 	if err := s.Delete(ctx, srcID); err != nil {
 		return nil, err
 	}
 
+	log.C(ctx).Infof("Deleting source labels for application with id %s", srcID)
+	if err := s.labelRepo.DeleteAll(ctx, appTenant, model.ApplicationLabelableObject, srcID); err != nil {
+		return nil, err
+	}
+
+	log.C(ctx).Infof("Updating destination app with id %s", srcID)
 	if err := s.appRepo.Update(ctx, appTenant, destApp); err != nil {
 		return nil, err
 	}
 
+	for _, destAppLabel := range destAppLabelsMerged {
+		log.C(ctx).Infof("Upserting destination label with id %s", destAppLabel.ID)
+		if err := s.labelRepo.Upsert(ctx, appTenant, destAppLabel); err != nil {
+			return nil, err
+		}
+	}
+
 	return destApp, nil
 }
 
+// handleMergeLabels merges source labels into destination labels
+func (s *service) handleMergeLabels(srcAppLabels, destAppLabels map[string]*model.Label) (map[string]*model.Label, error) {
+	destIntSysID := fmt.Sprintf("%v", destAppLabels[intSysKey].Value)
+	srcIntSysID := fmt.Sprintf("%v", srcAppLabels[intSysKey].Value)
+	if len(destIntSysID) == 0 && len(srcIntSysID) > 0 {
```

Comment:
Not sure that this case is even possible, but maybe it's a good sanity check

Reply:
Explanation here: https://github.com/kyma-incubator/compass/pull/2318#discussion_r870238316

Reply:
Yes, I just meant that since we have a sanity check beforehand that the `src` and `dest` have one and the same app template, we usually specify one and the same int system (when there is one) when an app template has a corresponding int system. So we kind of have the guarantee that if the app templates match and there is an int system associated with the app template, then both `src` and `dest` should have one and the same value for this label.

I don't see a scenario where two systems from one and the same app template have two different int sys IDs. So either they are the same, or it's missing (and so it's fine for the mergo lib to skip it or whatever it does when the value is empty string). I'm kind of iffy on whether we should keep this code, because practically it should not reach such a situation, theoretically everything is possible, but the trade off is that we have some cryptic incentive with this code part which people who aren't aware of this discussion here will wonder why we have this custom logic.